README.rst

```diff
@@ -1962,6 +1968,341 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
+We hope to add more types of macros in the future.
+
+Macro registry
+~~~~~~~~~~~~~~
+
+The macro registry is where the definitions of all currently defined macros are stored,
+together with the resulting GraphQL schema they form, as well as any associated metadata
+that the compiler's macro system may need to expand any macros encountered in a query.
+
+To create a macro registry object for a given GraphQL schema, use the :code:`create_macro_registry`
+function:
+
+.. code:: python
+
+    from graphql_compiler.macros import create_macro_registry
+
+    macro_registry = create_macro_registry(your_graphql_schema_object)
+
+To retrieve the GraphQL schema object with all its macro-based additions, use
+the :code:`get_schema_with_macros` function:
+
+.. code:: python
+
+    from graphql_compiler.macros import get_schema_with_macros
+
+    graphql_schema = get_schema_with_macros(macro_registry)
+
+Schema for defining macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Macro definitions rely on additional directives that are not normally defined in the schema
+the GraphQL compiler uses for querying. We intentionally do not include these directives in
+that schema, since defining macros and writing queries are different modes of use of the compiler,
+and we believe that controlling which sets of directives are available in which mode will minimize
+the potential for user confusion.
+
+You may use the :code:`get_schema_for_macro_definition()` function to transform a querying schema
+into one that is suitable for defining macros.
+
+Macro edges
+~~~~~~~~~~~
+
+Macro edges allow users to define new edges that become part of the GraphQL schema, using existing
+edges as building blocks. They allow users to define shorthand for common querying operations,
+encapsulating uses of existing query functionality (e.g., tags, filters, recursion,
+type coercions, etc.) into a virtual edge with a user-specified name that exists only on a specific
+GraphQL type (and all its subtypes). Both macro edge definitions and their uses are
+fully type-checked, ensuring the soundness of both the macro definition and any queries that use it.
+
+Overview and use of macro edges
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let us explain the idea of macro edges through a simple example.
+
+Consider the following query, which returns the list of grandchildren of a given animal:
+
+.. code ::
+
+    {
+        Animal {
+            name @filter(op_name: "=", value: ["$animal_name"])
+            out_Animal_ParentOf {
+                out_Animal_ParentOf {
+                    name @output(out_name: "grandchild_name")
+                }
+            }
+        }
+    }
+
+If operations on animals' grandchildren are common in our use case, we may wish that
+an edge like :code:`out_Animal_GrandparentOf` had existed and saved us some repetitive typing.
+
+One of our options is to materialize such an edge in the underlying database itself.
```

Comment:
I think that a lot of this text belongs in the motivation section I talked about earlier.